src/common/dependency-manager/local-manager.ts

```diff
@@ -119,72 +82,69 @@ export default class LocalDependencyManager extends DependencyManager {
-        const edge = new ServiceEdge(parent_node.ref, dep_node.ref);
-        this.graph.addEdge(edge);
+  async loadService(service_name: string, service_tag: string, recursive = true): Promise<ServiceNode | ExternalNode> {
+    const ref = `${service_name}:${service_tag}`;
+    const existing_node = this.graph.nodes_map.get(ref);
+    if (existing_node) {
+      return existing_node as ServiceNode | ExternalNode;
+    }
 
-        if (recursive) {
-          await this.loadDatastores(dep_node);
-          dependency_resolvers.push(() => this.loadDependencies(dep_node));
-        }
-      }
+    const env_service = this._environment.getServiceDetails(ref);
+    if (env_service?.getInterfaces() && Object.values(env_service?.getInterfaces()).every((i) => (i.host))) {
+      return this.loadExternalService(env_service, ref);
     }
 
-    await Promise.all(dependency_resolvers.map(fn => fn()));
-  }
+    const debug_path = env_service?.getDebugOptions()?.path;
+    let service_node;
+    if (debug_path) {
+      const svc_path = path.join(path.dirname(this.config_path), debug_path);
+      service_node = await this.loadLocalService(svc_path);
+    } else if (this.linked_services.hasOwnProperty(service_name)) {
+      console.log(`Using locally linked ${chalk.blue(service_name)} found at ${chalk.blue(this.linked_services[service_name])}`);
+      service_node = await this.loadLocalService(this.linked_services[service_name]);
+    } else {
+      const [account_name, svc_name] = service_name.split('/');
+      const { data: service_digest } = await this.api.get(`/accounts/${account_name}/services/${svc_name}/versions/${service_tag}`);
+
+      const service_config = ServiceConfigBuilder.buildFromJSON(service_digest.config);
+      service_node = new ServiceNode({
+        service_config: service_config,
+        node_config: this.getNodeConfig(service_config),
+        tag: service_digest.tag,
+        image: service_digest.service.url.replace(/(^\w+:|^)\/\//, ''),
+        digest: service_digest.digest,
+      });
+    }
 
-  /**
-   * @override
-   */
-  async loadService(service_name: string, service_tag: string): Promise<ServiceNode> {
-    const [account_name, svc_name] = service_name.split('/');
-    const { data: service_digest } = await this.api.get(`/accounts/${account_name}/services/${svc_name}/versions/${service_tag}`);
-
-    const config = ServiceConfigBuilder.buildFromJSON(service_digest.config);
-    const node = new ServiceNode({
-      service_config: config,
-      tag: service_digest.tag,
-      image: service_digest.service.url.replace(/(^\w+:|^)\/\//, ''),
-      ports: {
-        target: config.getPort() || 8080,
-        expose: await this.getServicePort(),
-      },
-      parameters: await this.getParamValues(
-        `${config.getName()}:${service_digest.tag}`,
-        config.getParameters(),
-      ),
-    });
-    this.graph.addNode(node);
-    return node;
+    this.graph.addNode(service_node);
```

Comment:
I know you didn't change this, but this was the reason I don't like the object-method pattern that we were talking about on Friday. If I call `loadService()` I expect to get a service back. I hardly expect that the method is _also_ going to modify some other state.

I think moving this would require some substantial rework (and probably is made more difficult by the recursive nature of this method), so I think it's okay to leave it... I just want to point out that this is a pattern I generally try to avoid since it leads to behavior that may not be expected by the caller.

Reply:
Fair enough - I'll see how heavy a lift it would be.